src/interfaces.ts

```diff
@@ -162,3 +167,27 @@ export interface CommentFlag {
     affectedRange: Range;
     codes: number[] | null;
 }
+
+type ValidateHandler = (scope: Scope, files: (BrsFile | XmlFile)[], callables: CallableContainerMap) => void;
+
+export interface CompilerPlugin {
+    name: string;
+    beforePrepublish?: (files: FileObj[]) => void;
+    beforePublish?: (files: FileObj[]) => void;
+    afterPublish?: (files: FileObj[]) => void;
+    programCreated?: (program: Program) => void;
```

Comment:
I'd like ensure there are `before` and `after` for every single events. Plugin authors won't necessarily understand the full flow of the compiler, and they shouldn't need to know it. For example, there's nothing telling plugin authors that `beforePublish` happens after `beforePrepublish`. They would expect to see a `beforePrepublish` and `afterPrepublish`. This also better insulates our plugin system from unexpected issues if we were to add some additional step between `beforePrepublish` and `beforePublish`. Here are the events I think we are missing:

- afterPrepublish
- beforeProgramCreate
- afterScopeCreate
- beforeScopeDispose
- beforeFileSourceLoad
- beforeFileParse
- afterFileValidate

Reply:
Ok I tweaked a the events.

Reply:
Not sure I would say it's "really" wasteful, as these events should be essentially no-ops if no plugins are subscribing to them. However, as long as our documentation clearly outlines the execution order of the events, I can be ok with omitting `beforeFileValidate`.

Reply:
You updated the docs to show the order of the events, that's good enough. :)